src/viser/_message_api.py

```diff
@@ -834,3 +834,30 @@ def send_file_download(self, filename: str, content: bytes) -> None:
                 mime_type=mime_type,
             )
         )
+
+    def request_file_upload(self, mime_type: str) -> dict:
+        """Request a file upload from the client."""
+        self._queue(_messages.FileUploadRequest(mime_type=mime_type))
+        file_upload_sema = threading.Semaphore(value=0)
+        file_content = {
+            "content": "",
+            "filename": "",
+            "mime_type": "",
+            "filled": False,
+        }
+
+        def handle_file_upload(client_id: ClientId, message: _messages.FileUpload):
+            file_content["content"] = message.content
+            file_content["filename"] = message.filename
+            file_content["mime_type"] = message.mime_type
+            file_content["filled"] = True
+            file_upload_sema.release()
+
+        self._message_handler.register_handler(_messages.FileUpload, handle_file_upload)
+        while not file_content["filled"]:
+            time.sleep(0.1)
+            file_upload_sema.acquire()
+        self._message_handler.unregister_handler(
+            _messages.FileUpload, handle_file_upload
+        )
+        return file_content
```

Comment:
Two Qs:
- Do you think there could be use cases where we call this function but don't want to want to wait until the upload finishes?
- Is there some way to communicate that the file upload was cancelled?

I'm wondering if it would make sense to return some kind of promise/handle object, maybe a `FileUploadHandle` that looks something like:
```python
@dataclasses.dataclass(frozen=True)
class UploadedFile:
    filename: str
    mime_type: str
    content: bytes

class FileUploadHandle:
    def get_status(self) -> Literal["waiting", "cancelled", "ready"]:
        ...
    def get_uploaded_file(self) -> Optional[UploadedFile]:
        ...
    def wait_for_response(self) -> None:
        ...
```

Reply:
oh cool yeah I can do that

Reply:
communicating that the file upload was cancelled is really hard and not super well solved in javascript. I've tried some of the existing solutions and none of them seem to work atm

Reply:
I see -- that seems problematic, no? Especially with the current API you could get stuck in an infinite loop?

As another suggestion, maybe we could refactor the code so the file upload is a GUI element? eg: https://mantine.dev/core/file-input/

And then we could just have a callback for when files are uploaded.

Maybe API could be:
```
upload_handle = server.add_gui_file_upload("Camera Trajectory", [any other args])

@upload_handle.on_upload
def _(file: UploadedFile) -> None:
    # Do something with the uploaded file
```

Reply:
Right now when you cancel, it just sends an empty file so it doesn't infinite loop. I can test more thoroughly though. I can also look into the option you're talking about

Reply:
@THE-COB @brentyi , can we (instead of using semaphore) use a callback which will be called when the file is uploaded?